conformance/endpoint_slice.go

```diff
@@ -0,0 +1,114 @@
+				"the service) so assuming the MCS implementation does not use EndpointSlices. If so, you can specify %q in the "+
+				"Ginkgo label filter where appropriate to skip this test entirely.", EndpointSliceLabel))
+		}
+	})
+
+	Specify("Exporting a service should create an EndpointSlice in the service's namespace in each cluster with the "+
+		"required MCS labels. Unexporting should delete the EndpointSlice.", func() {
+		AddReportEntry(SpecRefReportEntry, "https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#using-endpointslice-objects-to-track-endpoints")
+
+		endpointSlices := make([]*discoveryv1.EndpointSlice, len(clients))
+
+		for i, client := range clients {
+			eps := t.awaitMCSEndpointSlice(&client)
+			Expect(eps).NotTo(BeNil(), reportNonConformant(fmt.Sprintf("the EndpointSlice was not found on cluster %d", i+1)))
+
+			endpointSlices[i] = eps
+
+			Expect(eps.Labels).To(HaveKeyWithValue(v1alpha1.LabelServiceName, t.helloService.Name),
+				reportNonConformant(fmt.Sprintf("the EndpointSlice does not contain the %q label referencing the service name",
+					v1alpha1.LabelServiceName)))
+
+			Expect(eps.Labels).To(HaveKey(v1alpha1.LabelSourceCluster),
+				reportNonConformant(fmt.Sprintf("the EndpointSlice does not contain the %q label", v1alpha1.LabelSourceCluster)))
+
+			Expect(eps.Labels).To(HaveKey(discoveryv1.LabelManagedBy),
+				reportNonConformant(fmt.Sprintf("the EndpointSlice does not contain the %q label", discoveryv1.LabelManagedBy)))
+			Expect(eps.Labels[discoveryv1.LabelManagedBy]).ToNot(Equal(K8sEndpointSliceManagedByName),
+				reportNonConformant(fmt.Sprintf("the EndpointSlice's %q label must not reference %q",
+					discoveryv1.LabelManagedBy, K8sEndpointSliceManagedByName)))
+		}
+
+		By("Unexporting the service")
+
+		t.deleteServiceExport(&clients[0])
+
+		for i, client := range clients {
+			Eventually(func() bool {
+				_, err := client.k8s.DiscoveryV1().EndpointSlices(t.namespace).Get(ctx, endpointSlices[i].Name, metav1.GetOptions{})
+				return apierrors.IsNotFound(err)
+			}, 20*time.Second, 100*time.Millisecond).Should(BeTrue(),
+				reportNonConformant(fmt.Sprintf("the EndpointSlice was not deleted on unexport from cluster %d", i+1)))
+		}
+	})
+})
+
+func (t *testDriver) awaitMCSEndpointSlice(c *clusterClients) *discoveryv1.EndpointSlice {
+	var endpointSlice *discoveryv1.EndpointSlice
+
+	_ = wait.PollUntilContextTimeout(ctx, 100*time.Millisecond,
+		20*time.Second, true, func(ctx context.Context) (bool, error) {
+			defer GinkgoRecover()
+
+			list, err := c.k8s.DiscoveryV1().EndpointSlices(t.namespace).List(ctx, metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred(), "Error retrieving EndpointSlices")
+
+			for i := range list.Items {
+				eps := &list.Items[i]
+
+				if slices.IndexFunc(eps.OwnerReferences, func(r metav1.OwnerReference) bool {
+					return r.Kind == "Service"
```

Comment:
That leads to a Catch-22 — one of the purposes of the test is to verify that EndpointSlices created by the MCS controller have the appropriate “managed by” label :wink:. Given the leniency of the spec though I suspect we won’t be able to come up with a guaranteed test. Perhaps checking that the “managed by” label _isn’t_ `endpointslice-controller.k8s.io` is good enough: if an MCS controller uses that label value, it’s asking for trouble anyway, so it’s fair game to ignore such EndpointSlices IMO; and we can specify that the conformance suite shouldn’t be run with non-default EndpointSlice controllers. Another option would be to allow the user to specify the “managed-by” value to check for.

Reply:
I was going by the original statement in the spec, "_EndpointSlices will have an owner reference to their associated ServiceImport_", but that's been since removed. I didn't check for the `endpointslice.kubernetes.io/managed-by` label here in order to verify this statement in the spec, "_These EndpointSlice objects will be marked as managed by the clusterset service controller, so that the endpoint slice controller doesn’t delete them_". But I guess checking that the `endpointslice.kubernetes.io/managed-by` label is not equal to `endpointslice-controller.k8s.io` is the only reliable way to distinguish a K8s `EndpointSlice` from an MCS one.

Reply:
@skitt Our comments crossed. It sounds like we concluded the same thing re: checking `endpointslice.kubernetes.io/managed-by` here.

Another option is to fail the test as non-conformant if we don't find an `EndpointSlice` not managed by `endpointslice-controller.k8s.io`. If the implementation doesn't support MCS `EndpointSlices` then the user can filter out the test via the `EndpointSlice` Ginkgo label. But the test is also labeled as `Optional` so, by default, it would effectively be treated as required which may be inconsistent.

Reply:
> That sounds about right that was also very similar to what I had in mind too

Just want to clarify that you're referring to the "_fail the test as non-conformant if we don't find an EndpointSlice not managed by endpointslice-controller.k8s.io_" option?

Reply:
This seems fine too, the problem with skipping the test however is that if we really want to have this tested in a CI for instance and that we don't end up creating the endpointslice because of some implementation bugs it might silently be skipping this test :(.

Reply:
Right, but apart from asking the user whether the implementation under test is supposed to create EndpointSlices or not, I don’t see how to distinguish between “no EndpointSlices because the implementation doesn’t use them” (which is valid since this is optional) and “no EndpointSlices because of a bug”. This remains a conformance test, not a complete integration test…

Reply:
To me if you want to manually check if the implementation is conformant having the test fails shouldn't matter that much because it's in the "optional" section in the report and it could even makes more sense depending on how you look at it.

But for automated testing it would be safer to not have something that tries to skip this, because if you want to be thorough you would then have to actually check that it was not skipped.

Although if you think the test should really be skipped we could very well keep this by default and have another arg/label to make it fail/signify that this is an automated run which could be done at a later time, it's not a big deal IMO.

Reply:
@skitt As I mentioned before we can let the user distinguish via the `EndpointSlice` Ginkgo label, ie if the implementation doesn’t use them then filter out with the label. The test would run by default and would fail if not supported but I think that's OK - we can indicate in the report message to use the label to filter.

Reply:
OK, so we’ll say that by default, the conformance suite checks for full conformance, and provide labels to use if users want to disable (optional) parts of the suite. Works for me!

Reply:
Ok - I changed the proposed search criteria to also look for any MCS label present so if the implementation did incorrectly set the `endpointslice.kubernetes.io/managed-by` label equal to `endpointslice-controller.k8s.io` but correctly set the MCS labels then we would accurately report that the `endpointslice.kubernetes.io/managed-by` label was not set correctly rather than just report that no MCS EPS was found.

Reply:
Sounds like Cilium is not fully compliant wrt "_These EndpointSlice objects will be marked as managed by the clusterset service controller, so that the endpoint slice controller doesn’t delete them._". I suppose we could relax this requirement in the spec.

Reply:
So who creates the local EPS, K8s or Cilium? If the former then I assume Cilium adds labels to a K8s EPS  - is that safe? If the latter, if still managed by K8s, couldn't it get deleted unexpectedly?

Reply:
I guess depending how you read this it might be fine with the current text but definitely ambiguous at best, I created this to try to relax that a bit: https://github.com/kubernetes/enhancements/pull/4930 :+1:

Reply:
> So who creates the local EPS, K8s or Cilium? If the former then I assume Cilium adds labels to a K8s EPS - is that safe? If the latter, if still managed by K8s, is this safe, ie couldn't it get deleted unexpectedly?

When we create the derived service we craft to add labels needed for the endpointslice and copy the selector if the service is exported from the local cluster. Which means that local endpoints are managed by the kubernetes controller directly while the remote ones we handle it in our controllers. This is a bit limited as it doesn't work with services whch doesn't have any selectors though.

Reply:
OK so the MCS labels get propagated to the local EPS by the K8s EPS controller via the service labels. In this case, Cilium is still compliant b/c the EPS isn't actually created/managed by an MCS controller although your https://github.com/kubernetes/enhancements/pull/4930/ PR further clarifies it in the spec.

So there's really no way to tell if the EPS was created by an MCS controller or the K8s controller, hence cannot verify the `endpointslice.kubernetes.io/managed-by` label. We could add a flag to try to guide the test as to who manages the EPS but Cilium muddies that further b/c some are MCS-managed and some not. So I think the only thing we can search for is the presence of at least one MCS label.

BTW, with https://github.com/kubernetes-sigs/mcs-api/pull/81 we could verify the value of the `multicluster.kubernetes.io/source-cluster` label although I'm not entirely sure if the cluster name extracted from the kubeconfig is guaranteed to match what the MCS implementation views as the cluster name.

Reply:
> OK so the MCS labels get propagated to the local EPS by the K8s EPS controller via the service labels. In this case, Cilium is still compliant b/c the EPS isn't actually created/managed by an MCS controller although your https://github.com/kubernetes/enhancements/pull/4930 PR further clarifies it in the spec.

> So there's really no way to tell if the EPS was created by an MCS controller or the K8s controller, hence cannot verify the endpointslice.kubernetes.io/managed-by label. We could add a flag to try to guide the test as to who manages the EPS but Cilium muddies that further b/c some are MCS-managed and some not. So I think the only thing we can search for is the presence of at least one MCS label.

Yeah I am thinking that the managed-by check could be an opt-out check somehow. IMO this should be fine as is, Cilium is probably the only one doing that and to integrate this test to Cilium correctly I would also have to  change a bit the test service because endpointslice syncing is opt-in in Cilium so I would probably need to add a way to have a way to hook into the Service creation or some kind of argument . This means that I could also add something to optionally ignore the managed-by check at the ~ same time.

> BTW, with https://github.com/kubernetes-sigs/mcs-api/pull/81 we could verify the value of the multicluster.kubernetes.io/source-cluster label although I'm not entirely sure if the cluster name extracted from the kubeconfig is guaranteed to match what the MCS implementation views as the cluster name (@skitt ).

What's in kubeconfig is kind of free form so I would guess that you could provide a kubeconfig that is different on many implementations x). Although I am guessing the conformance test could just assume that the conformance test make them match :man_shrugging:.

Reply:
> IMO this should be fine as is

I assume you mean the way it is now in my latest commit even tho it would fail on Cilium. Of course, you could always filter out this test since it's optional.

>  endpointslice syncing is opt-in in Cilium

We have a similar situation with Submariner re: `ServiceImport` VIP. By default we don't set this so the test fails. However we recently added an option to set the VIP so when running the conformance tests we set the option. However this causes the DNS test to fail as we rely on some external component to handle that. But the DNS test is optional so I can filter it out 😄 I assume you could do similar, ie filter out the EPS test according to your opt-in setting.

> add something to optionally ignore the managed-by check at the ~ same time.

I could do that with this PR.

Reply:
> > BTW, with #81 we could verify the value of the multicluster.kubernetes.io/source-cluster label although I'm not entirely sure if the cluster name extracted from the kubeconfig is guaranteed to match what the MCS implementation views as the cluster name (@skitt ).
> 
> What's in kubeconfig is kind of free form so I would guess that you could provide a kubeconfig that is different on many implementations x). Although the conformance test could probably just assume that the user provides something that actually match if it's clearly stated somewhere 🤷‍♂️.

Yup, the cluster name in the kubeconfig is only relevant on the client, it doesn’t have to be related to the cluster id used by the MCS controller.

Reply:
I added a `skip-verify-eps-managed-by` CLI flag defaulting to false.